proof/refine/ARM/Syscall_R.thy

```diff
@@ -858,6 +858,171 @@ lemma threadSet_not_curthread_ct_domain:
+  apply (rule hoare_seq_ext_skip, wpsimp)
+  apply (rule hoare_when_cases, simp)
+  apply (rule hoare_seq_ext[OF _ get_sc_sp'])
+  apply (wpsimp wp: set_sc'.obj_at' simp: updateSchedContext_def)
+  apply (clarsimp simp: obj_at'_real_def ko_wp_at'_def split: if_split)
+  done
+
+lemma schedContextUpdateConsumed_bound_scTCB[wp]:
+  "schedContextUpdateConsumed tptr \<lbrace>obj_at' (\<lambda>a. \<exists>y. scTCB a = Some y) scPtr\<rbrace>"
+  apply (clarsimp simp: schedContextUpdateConsumed_def)
+  apply (wpsimp wp: set_sc'.obj_at' simp: updateSchedContext_def)
+  apply (clarsimp simp: obj_at'_real_def ko_wp_at'_def split: if_split)
+  done
+
+crunches schedContextCompleteYieldTo
+  for bound_scTCB[wp]: "obj_at' (\<lambda>a. \<exists>y. scTCB a = Some y) scPtr"
+
+lemma contextYieldToUpdateQueues_invs':
+  "\<lbrace>invs' and (\<lambda>s. obj_at' (\<lambda>a. \<exists>y. scTCB a = Some y) scPtr s) and ct_active'
+    and ex_nonz_cap_to' scPtr and (\<lambda>s. tcb_at' (ksCurThread s) s)\<rbrace>
+   contextYieldToUpdateQueues scPtr
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  apply (clarsimp simp: contextYieldToUpdateQueues_def)
+  apply (rule hoare_seq_ext[OF _ get_sc_sp'], rename_tac sc)
+  apply (rule hoare_seq_ext[OF _ isSchedulable_sp])
+  apply (rule hoare_if; (solves wpsimp)?)
+  apply (rule hoare_seq_ext[OF _ getCurThread_sp], rename_tac ctPtr)
+  apply (rule hoare_seq_ext_skip, solves wpsimp)+
+  apply (rule hoare_if)
+   apply wpsimp
+   apply (erule isSchedulable_bool_runnableE)
+   apply (frule sc_ko_at_valid_objs_valid_sc')
+    apply fastforce
+   apply (clarsimp simp: valid_sched_context'_def valid_bound_obj'_def obj_at_simps opt_map_def)
+  apply (subst bind_dummy_ret_val[symmetric])
+  apply (subst bind_assoc[symmetric])
+  apply (rule_tac B="\<lambda>_. invs' and ct_active' and (\<lambda>s. st_tcb_at' runnable' (the (scTCB sc)) s)
+                         and (\<lambda>s. ctPtr = ksCurThread s)"
+               in hoare_seq_ext[rotated])
+   apply (clarsimp simp: pred_conj_def)
+   apply (intro hoare_vcg_conj_lift_pre_fix)
+       apply (rule hoare_weaken_pre)
+        apply (rule contextYieldToUpdateQueues_invs'_helper)
+       apply (fastforce dest: sc_ko_at_valid_objs_valid_sc'
+                        simp: valid_sched_context'_def valid_sched_context_size'_def)
+      apply (wpsimp wp: threadSet_ct_in_state' setSchedContext_ct_in_state')
+     apply (wpsimp wp: threadSet_st_tcb_at2)
+     apply (erule isSchedulable_bool_runnableE)
+     apply (frule sc_ko_at_valid_objs_valid_sc')
+      apply fastforce
+     apply (frule sc_ko_at_valid_objs_valid_sc')
+      apply fastforce
+     apply (clarsimp simp: valid_sched_context'_def scBits_simps obj_at_simps)
+    apply (wpsimp | wps)+
+  apply (clarsimp simp: ct_in_state'_def st_tcb_at'_def obj_at_simps runnable_eq_active')
+  done
+
+lemma tcbInReleaseQueue_update_st_tcb_at'[wp]:
+  "threadSet (tcbInReleaseQueue_update b) t \<lbrace>\<lambda>s. Q (st_tcb_at' P t' s)\<rbrace>"
```

Comment:
Probably unimportant if this hasn't been needed before, but maybe move this to TcbAcc_R?

Reply:
Some of the following more general lemmas could be moved earlier as well.